### PR TITLE
Captions reengineering.

### PR DESCRIPTION
**Pull request by diegoreymendez**
### Description:

This PR re-engineers captions in order to support a wider range of caption options (they could potentially be used for more than images now).

The old architecture was also causing some trouble when combining links with captions, which this PR will allow us to solve more elegantly.

To see the problem, simply go to `develop` and try setting a link for a captioned image.  The link is applied to the caption text too and there's no easy fix for it.

### Additional details:

Captions are now editable inline too, which is a good thing.

Captions now support links!

### Demo:

**Updated:**

captions

### Testing:

Just try to break captions in whatever way you deem possible.  Don't hesitate to take your time to test this PR, as it includes important architectural changes to the way captions work.

### CODE REVIEW

**Comment by diegoreymendez** [comment]
Thanks for the feedback!

File: Aztec/Classes/Extensions/NSAttributedString+Lists.swift [hunk 1]
```diff
@@ -5,6 +5,12 @@ import UIKit
 // MARK: - NSAttributedString Lists Helpers
 //
 extension NSAttributedString {
+    
+    /// Internal convenience helper. Returns the internal string as a NSString instance
+    ///
+    var foundationString: NSString {
```
[the remark this answers is not in the document]
Comment (diegoreymendez) [reply Aztec/Classes/Extensions/NSAttributedString+Lists.swift]: I changed the file's name.  It's showing up as new but it's only a rename.

[the remark this answers is not in the document]
Comment (SergioEstevao) [reply Aztec/Classes/Extensions/NSAttributedString+Lists.swift]: @jleandroperez if you look a bit lower down in the PR you will see that Diego is just moving this around.

File: Example/Example/EditorDemoController.swift [hunk 2]
```diff
@@ -1315,20 +1315,20 @@ private extension EditorDemoController
...
-           oldURL = url
-           detailsViewController.linkURL = url
+        
+        let caption = richTextView.caption(for: attachment)
+        let detailsViewController = AttachmentDetailsViewController.controller(for: attachment, with: caption)
+        
+        let linkInfo = richTextView.linkInfo(for: attachment)
+        let linkRange = linkInfo?.range
+        let linkURL = linkInfo?.url
+        let linkUpdateRange = linkRange ?? richTextView.textStorage.ranges(forAttachment: attachment).first!
```
Comment (SergioEstevao) [inline Example/Example/EditorDemoController.swift]: Are we sure we want to use the ! on this line?

**Comment by SergioEstevao** [comment]
Look like the UI Tests are failing. Not sure if this PR is to blame for that.

**Comment by diegoreymendez** [comment]
The UI tests have been failing on and off, in a not very consistent manner.  It seems like they're very sensitive to not only change, but also how the simulator is set up.

I was planning (if we agree) to tackle that separately because I've been seeing that behavior event before this PR.

**Comment by SergioEstevao** [comment]
While in general this PR is working very well I want to raise the following issues:

- You can delete an image attachment, and the caption stays there. Calypso deletes everything.
- When you do that we get `<figure>` and `<figcaption>` elements orphan on the HTML view.

**Comment by SergioEstevao** [comment]
Another issue that I found is the presentation of caption, it looks we are adding a lot of paragraph spacing between the image and the caption.

This make it look that the caption is not attached to the image.

**Comment by diegoreymendez** [comment]
@jleandroperez - Can you try again?  I found an issue with some paragraph-separators that were being lost on conversion.

**Comment by diegoreymendez** [comment]
@SergioEstevao - I'll try to reduce the spacing and add a specific style to captions.  Please hold tight.

**Comment by diegoreymendez** [comment]
@SergioEstevao - This PR now styles captions differently.  Also when pressing enter in the middle of a caption, the style will be removed for the right-hand text.

@SergioEstevao, @jleandroperez - Can I ask you for another run?

**Comment by diegoreymendez** [comment]
PS: right now the caption styling is defined by `FigcaptionFormatter`.  I'm planning to make it easier in an upcoming PR for the client App to customize this.

**Comment by SergioEstevao** [comment]
@diegoreymendez from my tests, it looks that now if you add a link to an image `<a>` it breaks the `[caption]` processor and you end up with `figure` and `figcaption` elements on the final HTML.

**Comment by diegoreymendez** [comment]
Very nice find, changes are up and ready for review.

**Comment by diegoreymendez** [comment]
Thanks @SergioEstevao !

**Pull request merged**